### PR TITLE
Improve code quality in History (complexity)

### PR DESCRIPTION
**Pull request by Lesterpig**
https://github.com/privly/privly-applications/issues/155

### CODE REVIEW

File: History/js/new.js [hunk 1]
```diff
@@ -149,85 +149,14 @@ var callbacks = {
...
     var tableBody = document.getElementById("table_body");
     for(var i = 0; i < response.json.length; i++) {
-
-      var href = response.json[i].privly_URL;
-      var app = response.json[i].privly_application;
-
-      // Rename deperecated apps
-      if ( app === "ZeroBin" ) {
-        app = "Message";
-      }
```
Comment (irdan) [inline History/js/new.js]: This if block still needs to exist in the refactored version.

File: History/js/new.js [hunk 2]
```diff
@@ -240,6 +169,80 @@ var callbacks = {
 };
 
 /**
+ * Transform the row object into a DOM object for a single message
```
Comment (smcgregor) [inline History/js/new.js]: Suggestion:

`Transform the row object into a DOM object for a single message` -> `Create a node for insertion into the jquery datatables plugin using plain javascript. Avoiding jquery and datatables APIs for this task reduces the risk of the content server XSS.`

Also, please note that multiple applications get listed by the History application, not just the "Message" application.

File: History/js/new.js (hunk 1 above)
Comment (smcgregor) [inline History/js/new.js]: We can remove this in a few weeks, but we need it at the moment -- I just need to find time to write and test a migration server side.

**Comment by smcgregor** [comment]
I have not tested this locally yet so issues may yet arise, but other than the comments above I think this looks good.

**Comment by Lesterpig** [comment]
Thanks! I'll fix issues from your comments :)

**Comment by irdan** [comment]
I have a feeling that the complexity of this code has not decreased significantly.  I would sign up for CodeClimate and point to your branch to see how much the complexity was reduced.

**Comment by Lesterpig** [comment]
@irdan Just C -> B in that case. It's mainly a test to setup correctly my environment.

**Comment by smcgregor** [comment]
Code quality heuristics don't like having big blocks of code inside loops, which we often unintentionally ignored. Regardless, this should be easier to test.

Speaking of which, have you gotten to setting up Karma yet? How do the tests look on this?

**Comment by Lesterpig** [comment]
@smcgregor I haven't found any test case for this callback function. Indeed, all tests are successful according to Karma.

**Comment by smcgregor** [comment]
For reference, here is the current build running from my fork: https://travis-ci.org/smcgregor/privly-applications/builds/52995589

**Comment by Lesterpig** [comment]
https://github.com/privly/privly-applications/issues/157

File: History/test/new.js [hunk 3]
```diff
@@ -95,6 +95,48 @@ describe ("History Logged In New Suite", function() {
...
+    //Check content
+    var cells = jQuery("#table_body tr:first td");
+    var buttons = jQuery(cells[0]).children("button");
+    
+    expect(buttons[0].getAttribute("data-target")).toBe("#historyPreview");
+    expect(buttons[0].textContent).toMatch(/Preview Tester/);
+    expect(buttons[1].getAttribute("data-canonical-href")).toMatch(/show.html\?privlyOriginalURL\=testUrl/);
+
+    for(var i = 1; i <= 3; i++)
+      expect(cells[i].textContent).not.toBe(""); // TODO unit tests for parseDate
```
Comment (smcgregor) [inline History/test/new.js]: For better or for worse, I believe CodeClimate will complain about not bracketing this.

**Comment by smcgregor** [comment]
New unit test looks good. With one more for _exclusively_ `getMessageDOM`, I think you are ready to go.

File: History/test/new.js (hunk 3 above)
Comment (Lesterpig) [inline History/test/new.js]: Fixed!

**Comment by smcgregor** [comment]
Two tiny comments left.

Also, make sure you comment when you push new commits, otherwise people don't get emailed about it.

File: History/test/new.js [hunk 4]
```diff
@@ -135,3 +178,31 @@ describe ("History Logged out New Suite", function() {
...
+
+      // Check content
+      expect(jQuery(cells[0]).children("button")[0].textContent).toMatch(/Preview Tester/);
+      for(var i = 1; i <= 3; i++) {
+        expect(cells[i].textContent).not.toBe(""); // TODO unit tests for parseDate
+      }
+
+    });
+  });
+});
```
Comment (smcgregor) [inline History/test/new.js]: nit: newline

Comment (smcgregor) [inline History/test/new.js]: So many PRs to get through, didn't mean to repeat this.

**Comment by Lesterpig** [comment]
No problem! :)

**Comment by smcgregor** [comment]
I tested locally and all is good. I am just waiting for the tests to run on Continuous Integration, then I'll merge.

Note: You are likely going to have issues with your fork since you are developing on master. I recommend working on a feature branch.

**Pull request merged**